fem/quadinterpolator_face.cpp

```diff
@@ -355,6 +353,184 @@ void FaceQuadratureInterpolator::Eval3D(
    });
 }
 
+template<const int T_VDIM, const int T_ND1D, const int T_NQ1D>
+void FaceQuadratureInterpolator::SmemEval3D(
+   const int NF,
+   const int vdim,
+   const DofToQuad &maps,
+   const Array<bool> &signs,
+   const Vector &e_vec,
+   Vector &q_val,
+   Vector &q_der,
+   Vector &q_det,
+   Vector &q_nor,
+   const int eval_flags)
+{
+   MFEM_PERF_SCOPE("FaceQuadInterpolator::SmemEval3D");
+   const int nd1d = maps.ndof;
+   const int nq1d = maps.nqpt;
+   const int ND1D = T_ND1D ? T_ND1D : nd1d;
+   const int NQ1D = T_NQ1D ? T_NQ1D : nq1d;
+   const int VDIM = T_VDIM ? T_VDIM : vdim;
+   MFEM_VERIFY(ND1D <= MAX_ND1D, "");
+   MFEM_VERIFY(NQ1D <= MAX_NQ1D, "");
+   MFEM_VERIFY(VDIM == 3 || !(eval_flags & DETERMINANTS), "");
+   auto B = Reshape(maps.B.Read(), NQ1D, ND1D);
+   auto G = Reshape(maps.G.Read(), NQ1D, ND1D);
+   auto F = Reshape(e_vec.Read(), ND1D, ND1D, VDIM, NF);
+   auto sign = signs.Read();
+   auto val = Reshape(q_val.Write(), NQ1D, NQ1D, VDIM, NF);
+   // auto der = Reshape(q_der.Write(), NQ1D, VDIM, 3, NF);
+   auto det = Reshape(q_det.Write(), NQ1D, NQ1D, NF);
+   auto nor = Reshape(q_nor.Write(), NQ1D, NQ1D, 3, NF);
+   MFEM_VERIFY(eval_flags | DERIVATIVES,
+               "Derivatives on the faces are not yet supported.");
+
+   MFEM_FORALL_3D(f, NF, NQ1D, NQ1D, VDIM,
+   {
+      constexpr int max_ND1D = T_ND1D ? T_ND1D : MAX_ND1D;
+      constexpr int max_NQ1D = T_NQ1D ? T_NQ1D : MAX_NQ1D;
+      constexpr int max_VDIM = T_VDIM ? T_VDIM : MAX_VDIM3D;
+
+      MFEM_SHARED double sm1[max_NQ1D*max_NQ1D*max_VDIM];
+      MFEM_SHARED double sm2[max_NQ1D*max_ND1D*max_VDIM];
+
+      auto r_F = (double(*)[max_ND1D][max_VDIM])sm1;
+      MFEM_FOREACH_THREAD(d1,x,ND1D)
+      {
+         MFEM_FOREACH_THREAD(d2,y,ND1D)
+         {
+            MFEM_FOREACH_THREAD(c,z,VDIM)
+            {
+               r_F[d1][d2][c] = F(d1,d2,c,f);
+            }
+         }
+      }
+      MFEM_SYNC_THREAD;
+
+      if (eval_flags & VALUES)
+      {
+         auto Bu = (double (*)[max_ND1D][max_VDIM])sm2;
+         MFEM_FOREACH_THREAD(d2,x,ND1D)
+         {
+            MFEM_FOREACH_THREAD(q,y,NQ1D)
```

Comment:
For consistency with the following code:
```suggestion
            MFEM_FOREACH_THREAD(q1,y,NQ1D)
```